src/aws/flb_aws_error_reporter.c

```diff
@@ -0,0 +1,247 @@
+    char *value, *end;
+    long result;
+
+    value = getenv(name);
+    if (!value) {
+        return 0;
+    }
+
+    result = strtol(value, &end, 10);
+    if (*end != '\0') {
+        return 0;
+    }
+    return (int) result;
+}
+
+/* create an error reporter*/
+struct flb_aws_error_reporter *flb_aws_error_reporter_create()
+{
+    char *path_var = NULL;
+    int ttl_var, status_message_length;
+    struct flb_aws_error_reporter *error_reporter;
+
+    error_reporter = flb_calloc(1, sizeof(struct flb_aws_error_reporter));
+    if (!error_reporter) {
+        flb_errno();
+        return NULL;
+    }
+
+    /* setup error report file path */
+    path_var = getenv(STATUS_MESSAGE_FILE_PATH_ENV);
+    if (path_var == NULL) {
+        flb_free(error_reporter);
+        return NULL;
+    }
+    error_reporter->file_path = flb_sds_create(path_var);
+
+    /* clean up existing file*/
+    remove(error_reporter->file_path);
+
+    /* setup error reporter message TTL */
+    ttl_var = getenv_int(STATUS_MESSAGE_TTL_ENV);
+    if (ttl_var <= 0) {
+        ttl_var = STATUS_MESSAGE_TTL_DEFAULT;
+    }
+    error_reporter->ttl = ttl_var;
+
+    /* setup error reporter file size */
+    status_message_length = getenv_int(STATUS_MESSAGE_MAX_BYTE_LENGTH_ENV);
+    if(status_message_length <= 0) {
+        status_message_length = STATUS_MESSAGE_MAX_BYTE_LENGTH_DEFAULT;
+    }
+    error_reporter->max_size = status_message_length;
+
+    /* create the message Linked Lists */
+    mk_list_init(&error_reporter->messages);
+
+    return error_reporter;
+}
+
+/* error reporter write the error message into reporting file and memory*/
+int flb_aws_error_reporter_write(struct flb_aws_error_reporter *error_reporter, char *msg)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_error_message *message;
+    struct flb_error_message *tmp_message;
+    flb_sds_t buf;
+    flb_sds_t buf_tmp;
+    int deleted_message_count = 0;
+    FILE *f;
+
+    buf = flb_sds_create(msg);
+    if (!buf) {
+        flb_errno();
+        return -1;
+    }
+    /* check if the message is the same with latest one in queue*/
+    if (mk_list_is_empty(&error_reporter->messages) != 0) {
+        tmp_message = mk_list_entry_last(&error_reporter->messages,
+                                         struct flb_error_message, _head);
+        if (tmp_message->len == flb_sds_len(buf) &&
+                flb_sds_cmp(tmp_message->data, buf, tmp_message->len) == 0) {
+
+            tmp_message->timestamp = time(NULL);
+            flb_sds_destroy(buf);
+            return 0;
+        }
+    }
+
+    message = flb_malloc(sizeof(struct flb_error_message));
+    if (!message) {
+        flb_sds_destroy(buf);
+        flb_errno();
+        return -1;
+    }
+    /* check if new message is too large and truncate*/
+    if (flb_sds_len(buf) > error_reporter->max_size) {
+        // truncate message
+        buf_tmp = flb_sds_copy(buf, msg, error_reporter->max_size);
```

Comment:
I think flb_sds_copy can allocation memory and you have to check for failure

Reply:
updated the way to check failure

Reply:
Since flb_sds_copy can allocate memory, I think you also have to do this after the if block:

```
if (buf != buf_tmp) {
      buf = buf_tmp
}
```

Because if it allocated memory than the `buf` pointer would then be invalid.